### PR TITLE
Change candidates/kickoutList state DB implementation

### PR DESCRIPTION
**Pull request by koseoyoung**
We've been storing candidate information by height key, so I change to only store current/next epoch's candidate/kickout information. And the archive node will store all of it. 
+ one issue: how should we rename the funcs (~byepoch) and API not supporting byEpoch 

setCandidate -> store with nextKey (mid of epoch)
setKickout -> store with next Key (end of epoch) 

shift -> move next data to current data (start of epoch at createPreState()) 

getCandidates -> read with currentKey (in consensus, start of epoch, with next key) 
getKickouts -> read with currentKey (in consensus, start of epoch, with next key)

### CODE REVIEW

**Comment by koseoyoung** [comment]
latest commit is that 
1. poll protocol interface has 4 func- Delegates/ DelegatesOfNextEpoch/ Candidates/ CandidatesOfNextEpoch 
2. In side of consensus and blockchain, it compares tip<-> target and calls a corresponding func (call back func)
3. Because now retrieving only tipEpoch delegate and candidate data is available, fixed some dependency with consensus unit test/ API

File: action/protocol/poll/util.go [hunk 1]
```diff
@@ -170,17 +176,78 @@ func setCandidates(
...
+		return err
+	}
+	nextKey := candidatesutil.ConstructKey(candidatesutil.NxtCandidateKey)
+	_, err := sm.PutState(&candidates, protocol.KeyOption(nextKey[:]))
 	return err
 }
 
-// setKickoutBlackList sets the blacklist for kick-out for corresponding epoch
+// setKickoutBlackList sets the blacklist for kick-out with next key
 func setKickoutBlackList(
```
Comment (CoderZhi) [inline action/protocol/poll/util.go]: setNextEpochBlacklist

File: action/protocol/poll/util.go [hunk 2]
```diff
@@ -170,17 +176,78 @@ func setCandidates(
...
 	sm protocol.StateManager,
 	blackList *vote.Blacklist,
-	epochNum uint64,
 ) error {
-	blackListKey := candidatesutil.ConstructBlackListKey(epochNum)
-	_, err := sm.PutState(blackList, protocol.LegacyKeyOption(blackListKey))
+	blackListKey := candidatesutil.ConstructKey(candidatesutil.NxtKickoutKey)
+	_, err := sm.PutState(blackList, protocol.KeyOption(blackListKey[:]))
```
Comment (CoderZhi) [inline action/protocol/poll/util.go]: NamespaceOption?

File: action/protocol/poll/util.go [hunk 3]
```diff
@@ -170,17 +176,78 @@ func setCandidates(
...
 		)
 	}
-	_, err := sm.PutState(&candidates, protocol.LegacyKeyOption(candidatesutil.ConstructKey(height)))
+	hu := config.NewHeightUpgrade(&bcCtx.Genesis)
+	if hu.IsPre(config.Easter, height) {
+		_, err := sm.PutState(&candidates, protocol.LegacyKeyOption(candidatesutil.ConstructLegacyKey(height)))
+		return err
+	}
+	nextKey := candidatesutil.ConstructKey(candidatesutil.NxtCandidateKey)
+	_, err := sm.PutState(&candidates, protocol.KeyOption(nextKey[:]))
```
Comment (CoderZhi) [inline action/protocol/poll/util.go]: SystemNamespace

File: action/protocol/poll/util.go [hunk 4]
```diff
@@ -170,17 +176,78 @@ func setCandidates(
...
 	return err
 }
+
+// setUnproductiveDelegates sets the upd struct with updkey
+func setUnproductiveDelegates(
+	sm protocol.StateManager,
+	upd *vote.UnproductiveDelegate,
+) error {
+	updKey := candidatesutil.ConstructKey(candidatesutil.UnproductiveDelegateKey)
+	_, err := sm.PutState(upd, protocol.KeyOption(updKey[:]))
```
Comment (CoderZhi) [inline action/protocol/poll/util.go]: same here

File: action/protocol/poll/util.go [hunk 5]
```diff
@@ -170,17 +176,78 @@ func setCandidates(
...
+	nextKey := candidatesutil.ConstructKey(candidatesutil.NxtCandidateKey)
+	_, err := sm.State(&next, protocol.KeyOption(nextKey[:]))
+	if err != nil {
+		return errors.Wrap(
+			err,
+			"failed to read next blacklist when shifting to current blacklist",
+		)
+	}
+	curKey := candidatesutil.ConstructKey(candidatesutil.CurCandidateKey)
+	_, err = sm.PutState(&next, protocol.KeyOption(curKey[:]))
```
Comment (CoderZhi) [inline action/protocol/poll/util.go]: do a redundant checking that the height returned by sm.State and sm.PutState are identical

File: action/protocol/poll/util.go [hunk 6]
```diff
@@ -170,17 +176,78 @@ func setCandidates(
...
+	nextKey := candidatesutil.ConstructKey(candidatesutil.NxtKickoutKey)
+	_, err := sm.State(next, protocol.KeyOption(nextKey[:]))
+	if err != nil {
+		return errors.Wrap(
+			err,
+			"failed to read next blacklist when shifting to current blacklist",
+		)
+	}
+	curKey := candidatesutil.ConstructKey(candidatesutil.CurKickoutKey)
+	_, err = sm.PutState(next, protocol.KeyOption(curKey[:]))
```
Comment (CoderZhi) [inline action/protocol/poll/util.go]: same here

File: action/protocol/poll/util.go [hunk 7]
```diff
@@ -170,17 +176,78 @@ func setCandidates(
...
+	sm protocol.StateManager,
+	upd *vote.UnproductiveDelegate,
+) error {
+	updKey := candidatesutil.ConstructKey(candidatesutil.UnproductiveDelegateKey)
+	_, err := sm.PutState(upd, protocol.KeyOption(updKey[:]))
+	return err
+}
+
+// shiftCandidates updates current data with next data of candidate list
+func shiftCandidates(sm protocol.StateManager) error {
```
Comment (CoderZhi) [inline action/protocol/poll/util.go]: return (height, error)

File: action/protocol/poll/governance_protocol.go [hunk 8]
```diff
@@ -133,17 +142,25 @@ func (p *governanceChainCommitteeProtocol) CreatePreStates(ctx context.Context,
...
-		return setKickoutBlackList(sm, unqualifiedList, epochNum+1)
+		return setKickoutBlackList(sm, unqualifiedList)
+	} else if blkCtx.BlockHeight == epochStartHeight && hu.IsPost(config.Easter, epochStartHeight) {
+		if err := shiftCandidates(sm); err != nil {
+			return err
+		}
+		if blkCtx.BlockHeight == 1 {
+			return nil
+		}
+		return shiftKickoutList(sm)
```
Comment (CoderZhi) [inline action/protocol/poll/governance_protocol.go]: make sure the return height of shiftKickoutList and shiftCandidates are identical

File: action/protocol/poll/util.go [hunk 9]
```diff
@@ -170,17 +176,78 @@ func setCandidates(
...
+	zap.L().Debug("Shift candidatelist from next key to current key")
+	var next state.CandidateList
+	nextKey := candidatesutil.ConstructKey(candidatesutil.NxtCandidateKey)
+	_, err := sm.State(&next, protocol.KeyOption(nextKey[:]))
+	if err != nil {
```
Comment (CoderZhi) [inline action/protocol/poll/util.go]: one line

File: action/protocol/poll/util.go [hunk 10]
```diff
@@ -170,17 +176,78 @@ func setCandidates(
...
+		return errors.Wrap(
+			err,
+			"failed to read next blacklist when shifting to current blacklist",
+		)
+	}
+	curKey := candidatesutil.ConstructKey(candidatesutil.CurCandidateKey)
+	_, err = sm.PutState(&next, protocol.KeyOption(curKey[:]))
+	if err != nil {
```
Comment (CoderZhi) [inline action/protocol/poll/util.go]: one line

File: action/protocol/poll/util.go [hunk 11]
```diff
@@ -170,17 +176,78 @@ func setCandidates(
...
+	_, err := sm.State(next, protocol.KeyOption(nextKey[:]))
+	if err != nil {
+		return errors.Wrap(
+			err,
+			"failed to read next blacklist when shifting to current blacklist",
+		)
+	}
+	curKey := candidatesutil.ConstructKey(candidatesutil.CurKickoutKey)
+	_, err = sm.PutState(next, protocol.KeyOption(curKey[:]))
+	if err != nil {
```
Comment (CoderZhi) [inline action/protocol/poll/util.go]: one line

File: action/protocol/poll/governance_protocol.go [hunk 12]
```diff
@@ -152,15 +152,23 @@ func (p *governanceChainCommitteeProtocol) CreatePreStates(ctx context.Context,
 		if err != nil {
 			return err
 		}
-		return setKickoutBlackList(sm, unqualifiedList)
+		return setNextEpochBlacklist(sm, unqualifiedList)
 	} else if blkCtx.BlockHeight == epochStartHeight && hu.IsPost(config.Easter, epochStartHeight) {
```
Comment (CoderZhi) [inline action/protocol/poll/governance_protocol.go]: delete "else"

File: action/protocol/poll/governance_protocol.go [hunk 13]
```diff
@@ -152,15 +152,23 @@ func (p *governanceChainCommitteeProtocol) CreatePreStates(ctx context.Context,
...
-		if err := shiftCandidates(sm); err != nil {
+		prevHeight, err := shiftCandidates(sm)
+		if err != nil {
 			return err
 		}
 		if blkCtx.BlockHeight == 1 {
```
Comment (CoderZhi) [inline action/protocol/poll/governance_protocol.go]: move the logic into createGenesisStates.

Comment (koseoyoung) [reply action/protocol/poll/governance_protocol.go]: fixed in latest commit

File: action/protocol/poll/governance_protocol.go (hunk 12 above)
Comment (koseoyoung) [reply action/protocol/poll/governance_protocol.go]: fixed in latest commit

**Pull request merged**